### PR TITLE
[MM-57706] Calls: Participants list for call screen

### PR DESCRIPTION
**Pull request by cpoile**
#### Summary
- Add a participants list to the call screen
- Next up are the host controls user profile page
- @matthewbirtch I spent hours fiddling with the name field to make it ellipsis in the middle of the line. Either the Text control is bugged, or react native can't figure out the correct horizontal dimensions if it is a nested flexbox. The first screenshot is the closest I could get, but you can see that the resize calculator is ignoring the rest of the line (the microphone icons). So, if we want to have names ellipsized (which is important, imo) I think we're going to have to accept that the ellipsized text has to have a `flex: 1`, which means it will push the rest of the line to the right when it isn't ellipsized (i.e., the `(you)` and the `Host` badge will be right aligned).
- When there are too many participants to show in 80% expanded, the list will allow scrolling.
- Note: I don't have android or tablet because at the moment the build is broken and I can't seem to get it to boot up on Android without it crashing. Agniva messaged about it last week and I think Daniel is working on it.
- Designs: https://www.figma.com/file/8C0ojMc778Zx3a0nm66Edw/MM-56584-Host-Controls?node-id=1270%3A54747&mode=dev

| bugged ellipsis? | with 3 participants | with 17 default |
|---|---|---|
| IMG_1537 | IMG_1540 | IMG_1538 |
| with 17 expanded | the call screen | | |
 | IMG_1539 | IMG_1541 | | |

#### Ticket Link
- https://mattermost.atlassian.net/browse/MM-57706

#### Checklist

- [ ] Added or updated unit tests (required for all new features)
- [x] Has UI changes
- [x] Includes text changes and localization file updates
- [x] Have tested against the 5 core themes to ensure consistency between them.

#### Device Information
- iOS: 16.5.1, iPhone 14

#### Release Note
```release-note
Calls: Added a Participants list to the call screen
```

### CODE REVIEW

File: app/products/calls/screens/participants_list/pill.tsx [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+
+import {useTheme} from '@context/theme';
+import {changeOpacity} from '@utils/theme';
+import {typography} from '@utils/typography';
+
+type Props = {
+    text: string | number;
+}
+
+const getStyleSheet = ({theme}: { theme: Theme }) => {
```
Comment (enahum) [inline app/products/calls/screens/participants_list/pill.tsx:15]: please use `makeStyleSheetForTheme` instead of memoizing the style in the component

**Review by enahum — CHANGES_REQUESTED** [review]

File: .ruby-version [hunk 2]
```diff
@@ -1 +1 @@
-3.0.6
+3.2.2
```
[the remark this answers is not in the document]
Comment (cpoile) [reply .ruby-version:1]: accident, thanks

File: app/products/calls/screens/participants_list/participants_list.tsx [hunk 3]
```diff
@@ -0,0 +1,212 @@
...
+        },
+        picture: {
+            borderRadius: PROFILE_SIZE / 2,
+            height: PROFILE_SIZE,
+            width: PROFILE_SIZE,
+        },
+        name: {
+            ...typography('Body', 200, 'SemiBold'),
+            color: theme.centerChannelColor,
+            flex: 1,
```
[the remark this answers is not in the document]
Comment (cpoile) [reply app/products/calls/screens/participants_list/participants_list.tsx]: Believe me, I tried -- if you move the flex upwards then it turns out like the first screencap, and worse as you move it  further. 😓

File: app/products/calls/screens/participants_list/participants_list.tsx [hunk 4]
```diff
@@ -0,0 +1,212 @@
...
+    const intl = useIntl();
+    const theme = useTheme();
+    const currentCall = useCurrentCall();
+    const {bottom} = useSafeAreaInsets();
+    const {height} = useWindowDimensions();
+    const styles = useMemo(() => getStyleSheet({theme, bottom}), [theme, bottom]);
+    const isTablet = useIsTablet();
+    const Scroll = useMemo(() => (isTablet ? ScrollView : BottomSheetScrollView), [isTablet]);
+
+    const sessions = sortSessions(intl.locale, teammateNameDisplay, sessionsDict);
```
[the remark this answers is not in the document]
Comment (cpoile) [reply app/products/calls/screens/participants_list/participants_list.tsx]: Missed that, thanks.

File: app/products/calls/screens/participants_list/participants_list.tsx [hunk 5]
```diff
@@ -0,0 +1,212 @@
...
+                    <FormattedText
+                        style={styles.headerText}
+                        id={'mobile.calls_participants'}
+                        defaultMessage={'Participants'}
+                    />
+                    <Pill text={sessions.length}/>
+                </View>
+                <Scroll alwaysBounceVertical={false}>
+                    {sessions.map((sess) => (
+                        <Pressable
```
[the remark this answers is not in the document]
Comment (cpoile) [reply app/products/calls/screens/participants_list/participants_list.tsx]: Yep, with `FlatList`/`BottomSheetFlatList` extracting into a component makes sense.

File: app/products/calls/screens/participants_list/participant.tsx [hunk 6]
```diff
@@ -0,0 +1,159 @@
...
+                    size={PROFILE_SIZE}
+                    style={styles.profileIcon}
+                />
+            )}
+            <View style={styles.row}>
+                <Text
+                    style={styles.name}
+                    numberOfLines={1}
+                >
+                    {displayUsername(sess.userModel, intl.locale, teammateNameDisplay)}
```
Comment (enahum) [inline app/products/calls/screens/participants_list/participant.tsx:114]: `sess.userModel` won't update the details of the user model after fetched the first time, is that ok? do we expect to see the user display name changed real time if the user make changes to their profile? if the answer is yes, then we should observe the userModel for changes and if that is the case, same applies to the profile picture

**Review by enahum — APPROVED** [review]

**Review by streamer45 — APPROVED** [review]
Great work :+1:

File: app/products/calls/screens/participants_list/participant.tsx (hunk 6 above)
Comment (cpoile) [reply app/products/calls/screens/participants_list/participant.tsx:114]: Thanks, good catch. I've made the change here: https://github.com/mattermost/mattermost-mobile/pull/7905/commits/194cd7d609583235c2180ea94cf2b6bfc583f6c2
Removing the memo only causes the sort to rerun one more time (when a participant changes their name or profile picture), which is fine -- we already observe those few fields from their user model in the `observeCurrentSessionsDict` observer.

**Pull request merged**